### ISSUE TITLE
[salt-ssh] Salt-ssh Crashes Silently When Executed More Than Twice

### ISSUE THREAD

**Issue by srkunze** [saltstack/salt#27898]
In our setup, we currently run more than two salt-ssh simultaneously:

```
salt:/srv # salt-ssh minion state.sls host& 
[1] 13915
salt:/srv # salt-ssh minion state.sls host& 
[2] 13927
salt:/srv # salt-ssh minion state.sls host& 
[3] 13944
salt:/srv # salt-ssh minion state.sls host& 
[4] 13951
salt:/srv # salt-ssh minion state.sls host& 
[5] 13968

salt:/srv # minion:

Summary for minion
-----------
Succeeded: 0
Failed:   0
-----------
Total states run:    0
Total run time:  0.000 ms

minion:

Summary for minion
-----------
Succeeded: 0
Failed:   0
-----------
Total states run:    0
Total run time:  0.000 ms

minion:

Summary for minion
-----------
Succeeded: 0
Failed:   0
-----------
Total states run:    0
Total run time:  0.000 ms

minion:
----------
          ID: necessary_packages
    Function: pkg.installed
      Result: True
        [...]

minion:
----------
          ID: necessary_packages
    Function: pkg.installed
      Result: True
        [...]
```

Note that two salt-ssh calls in parallel work perfectly. More than 3 calls "fails" all of them except the last 2.

**Comment by jfindlay** [comment]
@srkunze, thanks for the report.

**Comment by srkunze** [comment]
Is there something I can do here?

**Comment by srkunze** [comment]
Still an issue with 2016.3.1:

```
salt:/srv # salt-ssh --versions-report
Salt Version:
           Salt: 2016.3.1
 
Dependency Versions:
           cffi: 0.8.6
       cherrypy: Not Installed
       dateutil: 2.1
          gitdb: 0.5.4
      gitpython: 0.3.2 RC1
          ioflo: Not Installed
         Jinja2: 2.8
        libgit2: 0.21.0
        libnacl: Not Installed
       M2Crypto: 0.22
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.7
   mysql-python: Not Installed
      pycparser: 2.10
       pycrypto: 2.6.1
         pygit2: 0.21.2
         Python: 2.7.3 (default, Apr 14 2012, 08:58:41) [GCC]
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.2.0
           RAET: Not Installed
          smmap: 0.8.2
        timelib: Not Installed
        Tornado: 4.3
            ZMQ: 4.1.2
 
System Versions:
           dist: SuSE 12.3 x86_64
        machine: x86_64
        release: 3.7.10-1.40-desktop
         system: Linux
        version: openSUSE  12.3 x86_64

```

**Comment by thatch45** [comment]
@srkunze sorry about the late reply here, have you tried running the commands with the `-W` flag?
This deploys the working code every time the command is run. Also, are you trying to execute multiple states at the same time? This will cause salt- to lock and prevent the state runs to run in parallel. the `-W` option will likely also allow you to run states like this since the locks will not be shared.

**Comment by srkunze** [comment]
@thatch45 Nope, we used the vanilla salt-ssh call. We actually want to prevent to transfer the thin-tgz every time for performance reasons.

> Also, are you trying to execute multiple states at the same time? 

Yes. Each salt-ssh run basically executes the same set of states (using different --pillar data). It basically creates different but similar users with a specific file and dir structure, services etc. IIRC, this behavior also happens for simple "cmd.run".

> the -W option will likely also allow you to run states like this since the locks will not be shared.

I think we can give it a try but actually I don't quite understand why the "0 states applies" output should happen at all (with or without the -W option).


**Currently, we "solve" the issue by "waiting" long enough:**

Say each salt-ssh call takes 5 minutes to complete. We wait 30 seconds between each call and there are 10 calls to be executed.

So, instead of waiting 5 minutes (starting all in parallel), we wait (10-1)*30 seconds + 5 minutes (for the last one to complete).

**Comment by thatch45** [comment]
This all makes sense, so I see a few issues. Once the initial thin is deployed I can run states in parallel, the main issue is the initial deploy of the thin. If you are trying to do that to the same location a bunch of times at the same time then this becomes a very complicated problem. Now, once that is out of the way is when we see the 0 states problem, this is because when we run states we ship a small tarball containing the state data and all the files that state data needs to execute, so what is likely happening is that you are placing the state data in the same location on multiple runs and only the last file write wins, so the other try to open up an incomplete state data tarball, read it as not having any states, and then return. The bug here is likely that the state call should return an error that it failed to read the tarball (if this is actually what happened)
again, the best way to sole this is with `-W`, now, as I write this, I do recall that we had added something that allows for the location of the thin on a redeploy basis to persist in a dedicated location. Let me look for that, it would allow for the behavior of -W without having to redeploy every time, instead each time you run in parallel you would get the same location as the last run.

**Comment by mephi42** [comment]
@thatch45 I can still see this problem with Salt 2017.7.3.

What do you think about randomizing `salt_state.tgz` name?
I made a quick patch and it seems to have worked.

The only thing is that we need to clean up those files now - maybe `state.pkg` can be made to do that?

**Comment by dmurphy18** [comment]
@srkunze Closing this due to age, the old version of Salt and Python 2.
Can you retest this with the latest release of Salt 3006.2 and if still an issue, please open a new issue, which will have metrics in tracking issues.
